drivers/hmis_external_apis/app/jobs/hmis_external_apis/ac_hmis/update_unit_availability_job.rb

```diff
@@ -8,36 +8,96 @@ module HmisExternalApis::AcHmis
   class UpdateUnitAvailabilityJob < ApplicationJob
     include HmisExternalApis::AcHmis::ReferralJobMixin
 
+    JOB_LOCK_NAME = 'hmis_external_update_unit_availability'.freeze
+
+    # @param data_source_id [Integer]
+    # @param force [Boolean]
+    def perform(data_source_id:, force: false)
+      with_locks do
+        projects = Hmis::Hud::Project.where(data_source_id: data_source_id)
+        projects.find_each do |project|
+          force_update(project) if force
+          project.external_unit_availability_syncs.dirty.preload(:unit_type, :user).each do |sync|
+            # sync.user made the most recent user changes, either to occupancy or unit inventory
+            sync_project_unit_type(
+              project: project,
+              unit_type: sync.unit_type,
+              user: sync.user || default_user,
+            )
+            # track sync version
+            sync.update!(synced_version: sync.local_version)
+          end
+        end
+      end
+    end
+
+    protected
+
+    def default_user
+      @default_user ||= Hmis::User.system_user
+    end
+
+    def force_update(project)
+      unit_types = project.units.preload(:unit_type).map(&:unit_type).compact.uniq
+      unit_types.each do |unit_type|
+        unit_type&.track_availability(project_id: project.id, user_id: default_user.id)
+      end
+    end
+
+    def with_locks
+      # lock job specific lock to prevent overlapping runs
+      Hmis::HmisBase.with_advisory_lock(JOB_LOCK_NAME, timeout_seconds: 0) do
+        # lock to synchronize with the project importer which may change available units
+        Hmis::HmisBase.with_advisory_lock(
+          HmisExternalApis::AcHmis::Importers::ProjectsImporter::JOB_LOCK_NAME,
+          timeout_seconds: 0,
+          shared: true,
+        ) do
+          yield
+          # FIXME ideally should report to dead man's snitch or equiv
+        end
+      end
+    end
+
     # determine current capacity at project and update external system
-    # @param project_id [Integer]
-    # @param unit_type_id [Integer]
-    # @param requested_by [String]
-    def perform(project_id:, unit_type_id:, requested_by:)
-      project = Hmis::Hud::Project.find(project_id)
+    # @param project [Hmis::Hud::Project]
+    # @param unit_type [Hmis::UnitType]
+    # @param user [Hmis::User]
+    def sync_project_unit_type(project:, unit_type:, user:)
       project_mper_id = mper.identify_source(project)
-      raise "mper id for not found Hmis::Hud::Project##{project_id}" unless project_mper_id
+      raise "mper id not found Hmis::Hud::Project##{project.id}" unless project_mper_id
```

Comment:
Rather than raising an error in this case, can we just ensure that the sync job ignores non-MPER projects? Or is that already happening? There will be one non-MPER project. It probably won't have unit types, but its possible that could change.

Reply:
Actually, nevermind. We dont really have a way to distinguish MPER/nonMPER programs. I didn't realize this line `project_mper_id = mper.identify_source(project)` is really just doing `project.ProjectID`.

I think what we want to do instead, is ignore the update if the _unit type_ does not have an external mper id. I think my comment addresses that: https://github.com/greenriver/hmis-warehouse/pull/3402/files#r1301598310

Reply:
I think you're right it is probably safer to skip records for unit types with missing mper ids. Otherwise one record will block further syncs. I changed it to log to sentry and continue. Seem okay? Also I removed the check for project_id since that will never be reachable, thanks for noticing that